gflownet/utils/batch.py

```diff
@@ -0,0 +1,346 @@
+                    self.envs[env_id].source,
+                    device=self.device,
+                    float_type=self.float,
+                )
+                # parent is not source
+                parents_policy[traj[1:]] = self.states_policy[traj[:-1]]
+                parents[traj[1:]] = self.states[traj[:-1]]
+            self.parents_policy = parents_policy
+            self.parents = parents
+
+    def merge(self, another_batch):
+        """
+        Merges two unprocessed batches
+        """
+        if self.is_processed or another_batch.is_processed:
+            raise Exception("Cannot merge processed batches")
+        if self.loss != another_batch.loss:
+            raise Exception("Cannot merge batches with different losses")
+        self.envs.update(another_batch.envs)
+        self.states += another_batch.states
+        self.actions += another_batch.actions
+        self.done += another_batch.done
+        self.env_ids += another_batch.env_ids
+        self.masks_invalid_actions_forward += (
+            another_batch.masks_invalid_actions_forward
+        )
+        self.masks_invalid_actions_backward += (
+            another_batch.masks_invalid_actions_backward
+        )
+        self.parents += another_batch.parents
+        self.all_possible_parents += another_batch.all_possible_parents
+        self.all_possible_parents_actions += another_batch.all_possible_parents_actions
+        self.steps += another_batch.steps
+
+    def _process_trajectory_indices(self):
+        """
+        Creates a dict of trajectory indices (key: env_id, value: indecies of the
+        states in self.states going in the order from s_1 to s_f. The dict is created
+        and stored in the self.trajectory_indices
+        """
+        trajs = defaultdict(list)
+        for idx, (env_id, step) in enumerate(zip(self.env_ids, self.steps)):
+            trajs[env_id.item()].append((idx, step))
+        trajs = {
+            env_id: list(map(lambda x: x[0], sorted(traj, key=lambda x: x[1])))
+            for env_id, traj in trajs.items()
+        }
+        self.trajectory_indices = trajs
+
+    def unpack_terminal_states(self):
+        """
+        For storing terminal states and trajectory actions in the buffer
+        Unpacks the terminating states and trajectories of a batch and converts them
+        to Python lists/tuples.
+        """
+        # TODO: make sure that unpacked states and trajs are sorted by traj_id (like
+        # rewards will be)
+        if not self.is_processed:
+            self.process_batch()
+        traj_actions = []
+        terminal_states = []
+        for traj_idx in self.trajectory_indices.values():
+            traj_actions.append(self.actions[traj_idx].tolist())
+            terminal_states.append(tuple(self.states[traj_idx[-1]].tolist()))
+        traj_actions = [tuple([tuple(a) for a in t]) for t in traj_actions]
+        return terminal_states, traj_actions
```

Comment:
Just a note that this will need to be rethought and re-implemented to adapt to current environments and needs. It is currently used only for the buffer and it seems it doesn't crash, so let's leave it as is for now. I have added TODO notes [here](https://github.com/alexhernandezgarcia/gflownet/blob/batch_class-ahg/gflownet/utils/batch.py#L375). To clarify, you haven't done anything wrong, Sasha! ;) (it's the old `unpack` method that was outdated).